File: 3/solution.erl
-module (solution).

main([Arg]) ->
  Number = list_to_integer(Arg),
  io:format("largest prime factor of number ~p is ~p ~n", [Number, lists:last(lists:sort(factors(Number)))]).
  
factors(Number) ->
  NotAFactor = fun(X) -> Number rem X /= 0 end,
  Factor = Number div lists:nth(1, lists:dropwhile(NotAFactor, lists:reverse(lists:seq(1,trunc(math:sqrt(Number)))))),
  
  case Number div Factor of
    TheOtherFactor when TheOtherFactor == 1; TheOtherFactor == Number ->
      [Number];
    TheOtherFactor ->
      factors(Factor) ++ factors(TheOtherFactor)
  end.
  

File: 3/solution.rb
def largest_prime_factor(n)
  factor = 2
  while factor < n
    until n % factor != 0
      n /= factor
    end
    if factor % 2 == 0
      factor += 1
    else
      factor += 2
    end
  end
  n
end

p largest_prime_factor(600851475143)